File: weatherroad/planner.py
"""Route planning engine that combines weather data with route information.

Analyzes weather conditions along route segments and provides safety
assessments and driving recommendations for Iceland travel.
"""

from dataclasses import dataclass, field
from enum import Enum

from .routes import RouteSegment, get_all_station_ids, get_route_segments, get_segment
from .weather import StationForecast, WeatherObservation, fetch_road_conditions, fetch_weather


class SafetyLevel(Enum):
    """Overall safety classification for a route segment."""

    SAFE = "safe"
    CAUTION = "caution"
    HAZARDOUS = "hazardous"
    DANGEROUS = "dangerous"


@dataclass
class SegmentAssessment:
    """Weather and safety assessment for a single route segment."""

    segment: RouteSegment
    safety_level: SafetyLevel
    weather_data: list[WeatherObservation] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)
    recommendations: list[str] = field(default_factory=list)
    road_condition: str = ""
    estimated_extra_time_minutes: int = 0


@dataclass
class RouteAssessment:
    """Complete assessment for a multi-segment route."""

    route_name: str
    segments: list[SegmentAssessment]
    overall_safety: SafetyLevel
    total_distance_km: float
    total_warnings: list[str] = field(default_factory=list)

    @property
    def total_extra_time_minutes(self) -> int:
        return sum(s.estimated_extra_time_minutes for s in self.segments)


def _assess_observation(obs: WeatherObservation) -> tuple[SafetyLevel, list[str], list[str]]:
    """Assess a single weather observation for driving safety."""
    warnings = []
    recommendations = []
    worst_level = SafetyLevel.SAFE

    # Wind assessment
    if obs.wind_speed is not None:
        if obs.wind_speed >= 25:
            worst_level = SafetyLevel.DANGEROUS
            warnings.append(
                f"Extreme wind: {obs.wind_speed} m/s at {obs.station_name}. "
                "High risk of vehicle being blown off road."
            )
            recommendations.append("Do NOT drive. Wait for conditions to improve.")
        elif obs.wind_speed >= 18:
            worst_level = max(worst_level, SafetyLevel.HAZARDOUS, key=lambda x: list(SafetyLevel).index(x))
            warnings.append(f"Very strong wind: {obs.wind_speed} m/s at {obs.station_name}")
            recommendations.append(
                "Only drive if necessary. Use extreme caution, especially on exposed roads."
            )
        elif obs.wind_speed >= 12:
            worst_level = max(worst_level, SafetyLevel.CAUTION, key=lambda x: list(SafetyLevel).index(x))
            warnings.append(f"Strong wind: {obs.wind_speed} m/s at {obs.station_name}")
            recommendations.append("Reduce speed and maintain firm grip on steering wheel.")

    # Wind gust assessment
    if obs.wind_gust is not None and obs.wind_gust >= 25:
        if worst_level.value in ("safe", "caution"):
            worst_level = SafetyLevel.HAZARDOUS
        warnings.append(f"Dangerous wind gusts: {obs.wind_gust} m/s at {obs.station_name}")

    # Visibility assessment
    if obs.visibility is not None:
        if obs.visibility < 500:
            worst_level = max(worst_level, SafetyLevel.DANGEROUS, key=lambda x: list(SafetyLevel).index(x))
            warnings.append(f"Near-zero visibility: {obs.visibility}m at {obs.station_name}")
            recommendations.append("Do NOT drive. Visibility dangerously low.")
        elif obs.visibility < 2000:
            worst_level = max(worst_level, SafetyLevel.HAZARDOUS, key=lambda x: list(SafetyLevel).index(x))
            warnings.append(f"Poor visibility: {obs.visibility}m at {obs.station_name}")
            recommendations.append("Use fog lights. Drive very slowly.")
        elif obs.visibility < 5000:
            worst_level = max(worst_level, SafetyLevel.CAUTION, key=lambda x: list(SafetyLevel).index(x))
            warnings.append(f"Reduced visibility: {obs.visibility}m at {obs.station_name}")

    # Temperature / icy road assessment
    if obs.temperature is not None and obs.temperature <= 0:
        if worst_level == SafetyLevel.SAFE:
            worst_level = SafetyLevel.CAUTION
        warnings.append(f"Freezing conditions: {obs.temperature}°C at {obs.station_name}")
        recommendations.append("Roads may be icy. Reduce speed and increase following distance.")

    if obs.road_temperature is not None and obs.road_temperature <= -2:
        worst_level = max(worst_level, SafetyLevel.HAZARDOUS, key=lambda x: list(SafetyLevel).index(x))
        warnings.append(f"Road surface freezing: {obs.road_temperature}°C at {obs.station_name}")
        recommendations.append("Black ice likely. Drive with extreme caution.")

    # Precipitation
    if obs.precipitation is not None and obs.precipitation > 5:
        if worst_level == SafetyLevel.SAFE:
            worst_level = SafetyLevel.CAUTION
        warnings.append(f"Heavy precipitation: {obs.precipitation}mm at {obs.station_name}")

    return worst_level, warnings, recommendations


def _estimate_extra_time(safety: SafetyLevel, distance_km: float) -> int:
    """Estimate additional travel time in minutes due to conditions."""
    # Assume base speed ~80 km/h on Route 1
    base_minutes = (distance_km / 80) * 60
    multipliers = {
        SafetyLevel.SAFE: 0,
        SafetyLevel.CAUTION: 0.2,
        SafetyLevel.HAZARDOUS: 0.5,
        SafetyLevel.DANGEROUS: 1.0,
    }
    return int(base_minutes * multipliers[safety])


def assess_segment(
    segment: RouteSegment,
    station_forecasts: dict[int, StationForecast],
    road_conditions: dict[str, dict],
) -> SegmentAssessment:
    """Assess a single route segment using available weather data."""
    all_warnings = []
    all_recommendations = []
    worst_safety = SafetyLevel.SAFE
    weather_data = []

    for station_id in segment.weather_station_ids:
        sf = station_forecasts.get(station_id)
        if not sf or not sf.forecasts:
            continue
        # Use the first (most recent/nearest) forecast
        obs = sf.forecasts[0]
        weather_data.append(obs)

        level, warnings, recs = _assess_observation(obs)
        all_warnings.extend(warnings)
        all_recommendations.extend(recs)
        if list(SafetyLevel).index(level) > list(SafetyLevel).index(worst_safety):
            worst_safety = level

    # Check road conditions
    road_cond = ""
    for name, cond in road_conditions.items():
        if segment.from_town.lower() in name.lower() or segment.to_town.lower() in name.lower():
            road_cond = cond.get("condition", "")
            desc = cond.get("description", "")
            if desc:
                all_warnings.append(f"Road condition ({name}): {desc}")
            break

    # Deduplicate recommendations
    seen = set()
    unique_recs = []
    for r in all_recommendations:
        if r not in seen:
            seen.add(r)
            unique_recs.append(r)

    return SegmentAssessment(
        segment=segment,
        safety_level=worst_safety,
        weather_data=weather_data,
        warnings=all_warnings,
        recommendations=unique_recs,
        road_condition=road_cond,
        estimated_extra_time_minutes=_estimate_extra_time(worst_safety, segment.distance_km),
    )


def plan_route(
    route_name: str | None = None,
    segment_ids: list[str] | None = None,
) -> RouteAssessment:
    """Plan a route and assess weather conditions along it.

    Args:
        route_name: Name of a predefined route (e.g., "ring-road").
        segment_ids: List of segment IDs for a custom route.

    Returns:
        RouteAssessment with safety levels and recommendations.
    """
    if route_name:
        segments = get_route_segments(route_name)
        display_name = route_name
    elif segment_ids:
        segments = [get_segment(sid) for sid in segment_ids]
        segments = [s for s in segments if s is not None]
        display_name = " -> ".join(s.name for s in segments)
    else:
        raise ValueError("Provide either route_name or segment_ids")

    if not segments:
        raise ValueError(f"No segments found for route: {route_name or segment_ids}")

    # Fetch weather for all stations along the route
    station_ids = get_all_station_ids(segments)
    forecasts_list = fetch_weather(station_ids, data_type="forec")
    station_forecasts = {sf.station_id: sf for sf in forecasts_list}

    # Fetch road conditions
    road_conditions = fetch_road_conditions()

    # Assess each segment
    assessments = []
    for segment in segments:
        assessment = assess_segment(segment, station_forecasts, road_conditions)
        assessments.append(assessment)

    # Overall safety is the worst of any segment
    overall = SafetyLevel.SAFE
    total_warnings = []
    for a in assessments:
        if list(SafetyLevel).index(a.safety_level) > list(SafetyLevel).index(overall):
            overall = a.safety_level
        total_warnings.extend(a.warnings)

    total_distance = sum(s.distance_km for s in segments)

    return RouteAssessment(
        route_name=display_name,
        segments=assessments,
        overall_safety=overall,
        total_distance_km=total_distance,
        total_warnings=total_warnings,
    )

File: docs/planner.js
/**
 * Route planning engine - assesses weather conditions for driving safety.
 */

const SAFETY = { SAFE: "safe", CAUTION: "caution", HAZARDOUS: "hazardous", DANGEROUS: "dangerous" };
const SAFETY_ORDER = [SAFETY.SAFE, SAFETY.CAUTION, SAFETY.HAZARDOUS, SAFETY.DANGEROUS];

function worstOf(a, b) {
  return SAFETY_ORDER.indexOf(a) >= SAFETY_ORDER.indexOf(b) ? a : b;
}

function assessObservation(obs) {
  let level = SAFETY.SAFE;
  const warnings = [];
  const recs = [];

  if (obs.wind_speed !== null) {
    if (obs.wind_speed >= 25) {
      level = SAFETY.DANGEROUS;
      warnings.push(`Extreme wind: ${obs.wind_speed} m/s at ${obs.station_name}. High risk of vehicle being blown off road.`);
      recs.push("Do NOT drive. Wait for conditions to improve.");
    } else if (obs.wind_speed >= 18) {
      level = worstOf(level, SAFETY.HAZARDOUS);
      warnings.push(`Very strong wind: ${obs.wind_speed} m/s at ${obs.station_name}`);
      recs.push("Only drive if necessary. Use extreme caution, especially on exposed roads.");
    } else if (obs.wind_speed >= 12) {
      level = worstOf(level, SAFETY.CAUTION);
      warnings.push(`Strong wind: ${obs.wind_speed} m/s at ${obs.station_name}`);
      recs.push("Reduce speed and maintain firm grip on steering wheel.");
    }
  }

  if (obs.wind_gust !== null && obs.wind_gust >= 25) {
    level = worstOf(level, SAFETY.HAZARDOUS);
    warnings.push(`Dangerous wind gusts: ${obs.wind_gust} m/s at ${obs.station_name}`);
  }

  if (obs.visibility !== null) {
    if (obs.visibility < 500) {
      level = worstOf(level, SAFETY.DANGEROUS);
      warnings.push(`Near-zero visibility: ${obs.visibility}m at ${obs.station_name}`);
      recs.push("Do NOT drive. Visibility dangerously low.");
    } else if (obs.visibility < 2000) {
      level = worstOf(level, SAFETY.HAZARDOUS);
      warnings.push(`Poor visibility: ${obs.visibility}m at ${obs.station_name}`);
      recs.push("Use fog lights. Drive very slowly.");
    } else if (obs.visibility < 5000) {
      level = worstOf(level, SAFETY.CAUTION);
      warnings.push(`Reduced visibility: ${obs.visibility}m at ${obs.station_name}`);
    }
  }

  if (obs.temperature !== null && obs.temperature <= 0) {
    level = worstOf(level, SAFETY.CAUTION);
    warnings.push(`Freezing conditions: ${obs.temperature}\u00B0C at ${obs.station_name}`);
    recs.push("Roads may be icy. Reduce speed and increase following distance.");
  }

  if (obs.road_temperature !== null && obs.road_temperature <= -2) {
    level = worstOf(level, SAFETY.HAZARDOUS);
    warnings.push(`Road surface freezing: ${obs.road_temperature}\u00B0C at ${obs.station_name}`);
    recs.push("Black ice likely. Drive with extreme caution.");
  }

  if (obs.precipitation !== null && obs.precipitation > 5) {
    level = worstOf(level, SAFETY.CAUTION);
    warnings.push(`Heavy precipitation: ${obs.precipitation}mm at ${obs.station_name}`);
  }

  return { level, warnings, recs };
}

function estimateExtraTime(safety, km) {
  const baseMin = (km / 80) * 60;
  const mult = { safe: 0, caution: 0.2, hazardous: 0.5, dangerous: 1.0 };
  return Math.round(baseMin * (mult[safety] || 0));
}

function assessSegment(segment, stationForecasts, roadConditions) {
  let worst = SAFETY.SAFE;
  const allWarnings = [];
  const allRecs = [];
  const weatherData = [];

  for (const sid of segment.stations) {
    const sf = stationForecasts[sid];
    if (!sf || !sf.forecasts.length) continue;
    const obs = sf.forecasts[0];
    weatherData.push(obs);

    const { level, warnings, recs } = assessObservation(obs);
    worst = worstOf(worst, level);
    allWarnings.push(...warnings);
    allRecs.push(...recs);
  }

  let roadCond = "";
  for (const [name, cond] of Object.entries(roadConditions)) {
    if (name.toLowerCase().includes(segment.from.toLowerCase()) ||
        name.toLowerCase().includes(segment.to.toLowerCase())) {
      roadCond = cond.condition || "";
      if (cond.description) allWarnings.push(`Road condition (${name}): ${cond.description}`);
      break;
    }
  }

  const uniqueRecs = [...new Set(allRecs)];

  return {
    segment,
    safety_level: worst,
    weather: weatherData,
    warnings: allWarnings,
    recommendations: uniqueRecs,
    road_condition: roadCond,
    extra_time_minutes: estimateExtraTime(worst, segment.km),
  };
}

async function planRoute(routeName, segmentIds) {
  let segments;
  let displayName;

  if (routeName) {
    segments = getRouteSegments(routeName);
    displayName = NAMED_ROUTES[routeName]?.name || routeName;
  } else {
    segments = segmentIds.map(getSegment).filter(Boolean);
    displayName = segments.map(s => s.name).join(" \u2192 ");
  }

  if (!segments.length) throw new Error("No segments found");

  const stationIds = getAllStationIds(segments);
  const [forecasts, roadConditions] = await Promise.all([
    fetchWeather(stationIds),
    fetchRoadConditions(),
  ]);

  const assessments = segments.map(seg => assessSegment(seg, forecasts, roadConditions));

  let overall = SAFETY.SAFE;
  const totalWarnings = [];
  for (const a of assessments) {
    overall = worstOf(overall, a.safety_level);
    totalWarnings.push(...a.warnings);
  }

  const totalKm = segments.reduce((sum, s) => sum + s.km, 0);
  const totalExtra = assessments.reduce((sum, a) => sum + a.extra_time_minutes, 0);

  return {
    route_name: displayName,
    overall_safety: overall,
    total_distance_km: totalKm,
    total_extra_time_minutes: totalExtra,
    total_warnings: totalWarnings,
    segments: assessments,
  };
}
